design-proposals/migration-target.md

```diff
@@ -0,0 +1,221 @@
+We are not going to alter by any mean the `spec` stanza of the VM or the VMI objects so future migrations or the node placement after a restart of the VM are not going to be affected by a `addedNodeSelector` set on a specific `VirtualMachineInstanceMigration` object.
+If the target pod fails to be started, the `VirtualMachineInstanceMigration` object will be marked as failed as it can already happen today for other reasons.
+The reason of the eventual failure will be reported back as it gets reported back today when a migration fails due to other reasons. 
+
+### How to propagate the additional constraint (the name of a target node for instance) to the target virt-launcher pod
+
+If set, the `addedNodeSelector` will be appended to the required nodeAffinity rules (`spec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution`) eventually set by the VM owner on the VM spec. 
+From a logical perspective, this is a pure **AND**:
+
+Nodes need to satisfy `addedNodeSelector` NodeSelector from the `VirtualMachineInstanceMigration` object **AND** the `NodeAffinity` constraints defined by the VM owner on the VM spec (and taints and tolerations).
+`addedNodeSelector` will adopt the well known `nodeSelector` grammar from k8s.
+It's a pretty flexible and standard API. It can address:
+- the simplest use case (migrating to a named nome identified by its node name):
+  ```yaml
+  addedNodeSelector:
+    nodeSelectorTerms:
+      - matchFields:
+        - key: metadata.name
+          operator: In
+          values:
+            - <nodeName>
+  ```
+- more complex scenario like migrating to a scheduler chosen host within a group of hosts satisfying an additional requirement:
+```yaml
+addedNodeSelector:
+  nodeSelectorTerms:
+    - matchExpressions:
+      - key: disktype
+        operator: In
+        values:
+            - ssd
+```
+
+From a coding perspective the additional logic required on the migration controller is pretty simple.
+Pseudocode:
+```go
+if migration.Spec.AddedNodeSelector != nil {
+    if templatePod.Spec.Affinity.NodeAffinity == nil {
+        templatePod.Spec.Affinity.NodeAffinity = &k8sv1.NodeAffinity{
+            RequiredDuringSchedulingIgnoredDuringExecution: migration.Spec.AddedNodeSelector,
+        }
+    } else if templatePod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+        templatePod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution = migration.Spec.AddedNodeSelector
+    } else {
+        mergedNodeAffinity := &k8sv1.NodeSelector{
+            NodeSelectorTerms: append(templatePod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms, migration.Spec.AddedNodeSelector.NodeSelectorTerms...),
+        }
+        templatePod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution = mergedNodeAffinity
+    }
+}
+```
```

Comment:
Blame on me,
I already pointed it out here: https://github.com/kubevirt/community/pull/320#issuecomment-2338326549
but then I somehow underestimated it.

[k8s doc](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity) clearly states:

> If you specify multiple terms in nodeSelectorTerms associated with nodeAffinity types, then the Pod can be scheduled onto a node if one of the specified terms can be satisfied (terms are ORed).
> 
> If you specify multiple expressions in a single matchExpressions field associated with a term in nodeSelectorTerms, then the Pod can be scheduled onto a node only if all the expressions are satisfied (expressions are ANDed).

So if we will simply append an additional `nodeSelectorTerm`, it will *ORed* producing wrong results.
Since we want it to be *ANDed* the only option is taking the expressions on the nodeSelectorTerm added to the VMIM and add it to all the existing  `nodeSelectorTerms` (which are already *ORed*!)